### ISSUE TITLE
Bundle upgrade hangs

### ISSUE THREAD

**Issue by DimaS86** [oleg-shilo/wixsharp#836]
Hi Oleg! I have a question about upgrading the bundle. I created my own customized bootstrapper, everything seems to be working fine, but now I want to implement the upgrading feature. So, I set the MajorUpgrade to default. I can see that the related bundle is detected, then planned and even gets executed (in the ExecutePackageBegin event I see it), but it never completes, it just hangs until I go to the task manager and kill the process. When the new package is planned, the ActionState is Install, not MajorUpgrade, meanwhile, when the related bundle is detected the operation is assigned to MajorUpgrade. I just don't know where should I look for a mistake. Could you please help me?

**Comment by DimaS86** [comment]
I tried something else. I tried to use MajorUpgradeStrategy instead and remove the old files on the AfterInstallInitialize step. And it actually does remove the old files first, then installs the new files, but the step it hangs is the last one when it tries to remove the old bundle registry entrance. And for some reason, it couldn't. If I remove it manually after I kill the hung installation process it is successfully removed. Am I doing something wrong or is there any workaround? Thanks in advance for your answer.

**Comment by oleg-shilo** [comment]
Sorry, I am out of my depth here. Cannot offer you any meaningful advice.

If I was facing this sort of problem I would try to see what debugging the BA would give me. But probably you have already done that...

**Comment by DimaS86** [comment]
Finally, I figured it out. It was my fault. I ran an exepackage in the installation sequence, made from the 3rd party exe file, and for some reason, it was causing troubles. It is offtopic, but I am having problems with BinaryFileAction as well.
`
new ManagedProject("MyProject", new Binary(new Id("MyFile_exe"), "MyFile.exe"),
new Dir(...
new BinaryFileAction("MyFile_exe", "-install", Return.check, When.Before, Step.InstallFinalize, WixSharp.Condition.NOT_Installed) { Execute = Execute.deferred }
...)`
What am I doing wrong? The result of it is that when I try to install this msi package it rollbacks the installation. Without BinaryFileAction everything is working correctly.
